### PR TITLE
Prefs [TASK][29129] Create DAO class

### PR DESCRIPTION
**Pull request by viet06061999**
## Related Tickets
- [Task#29129: Prefs [TASK][29129] Create DAO class](https://edu-redmine.sun-asterisk.vn/issues/29129)

## WHAT
Create DAO classes
## HOW
Create interface DAO and implement interface
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            contentValues.put(User.NAME, booking.name)
+            contentValues.put(User.PHONE, booking.phone)
+            contentValues.put(HotelBooking.IN_DATE, booking.checkinDate)
+            contentValues.put(HotelBooking.OUT_DATE, booking.checkoutDate)
+            contentValues.put(HotelBooking.NUM_ADULT, booking.numAdult)
+            contentValues.put(HotelBooking.NUM_CHILDREN, booking.numChildren)
+            database.insert(HotelBooking.TABLE_NAME, null, contentValues)
+            true
+        } catch (e: Exception) {
+            e.printStackTrace()
```
Comment (tuannm-1276) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]: handle Ex đi em. cả những nơi khác nữa

Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]: vâng ạ

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAO.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.sunasterisk.travelapp.data.source.local.dao
+
+import com.sunasterisk.travelapp.data.models.HotelBooking
+
+interface BookingDAO {
+    fun insertBooking(booking: HotelBooking): Boolean
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAO.kt:6]: insertBooking mà trả về boolean thì nghe vẻ k hợp lý lắm

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+class BookingDAOImpl private constructor(database: DatabaseHelper) : BookingDAO {
+    private val database = database.writableDatabase
+
+    override fun insertBooking(booking: HotelBooking): Boolean =
+        if (!checkBookingExist(booking)) {
+            database.insert(HotelBooking.TABLE_NAME, null, getValues(booking)) > 0
+        } else {
+            false
+        }
```
Comment (huytq-0932) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt:18]:
```suggestion
        if (checkBookingExist(booking)) return false 
        return database.insert(HotelBooking.TABLE_NAME, null, getValues(booking)) > 0
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    override fun getBookings(): List<HotelBooking> {
+        val cursor = database.query(HotelBooking.TABLE_NAME, null, null, null, null, null, null)
+        val bookings: MutableList<HotelBooking> = ArrayList()
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]: tại sao là ArrayList() mà k phải MutableList thế

File: app/src/main/java/com/sunasterisk/travelapp/data/models/Reservation.kt [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
     val phone: String,
     val name: String,
     var arrivalTime: String,
     var numAdult: Int,
-    var numChildrent: Int
+    var numChildren: Int
 ) {
     companion object {
         const val TABLE_NAME = "reservation"
         const val ID = "Reservation_Id"
```
Comment (tuannm-1276) [inline app/src/main/java/com/sunasterisk/travelapp/data/models/Reservation.kt]:
```suggestion
        const val ID = "reservation_id"
```
tương tự các thuộc tính còn lại

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt (hunk 3 above)
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]:
```suggestion
        if (booking.isExist()) return false
        return database.insert(HotelBooking.TABLE_NAME, null, getValues(booking)) > 0
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+    override fun deleteBooking(booking: HotelBooking): Boolean =
+        database.delete(
+            HotelBooking.TABLE_NAME,
+            "${HotelBooking.ID} = ?",
+            arrayOf(booking.id.toString())
+        ) > 0
+
+    private fun checkBookingExist(booking: HotelBooking): Boolean {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]:
```suggestion
    private fun HotelBooking.isExist(): Boolean {
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        } else {
+            false
+        }
+
+    override fun getBookings(): List<HotelBooking> {
+        val cursor = database.query(HotelBooking.TABLE_NAME, null, null, null, null, null, null)
+        val bookings: MutableList<HotelBooking> = ArrayList()
+        if (cursor.moveToFirst()) {
+            while (!cursor.isAfterLast) {
+                bookings.add(getBookingFromCursor(cursor))
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]:
```suggestion
                bookings.add(HotelBooking(cursor))
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt (hunk 4 above)
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]:
```suggestion
        val bookings = mutableListOf<HotelBooking>()
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                Hotel(id = hotelId),
+                email,
+                phone,
+                name,
+                checkinDate,
+                checkoutDate,
+                numAdult,
+                numChildrent
+            )
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]: viết hàm này dưới dạng constructor của HotelBooking 
thêm 1 constructor cho HotelBooking(cursor)

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAO.kt (hunk 2 above)
Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAO.kt:6]: chỗ này em chỉ để xem insert thành công hay thất bại thôi ạ

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt (hunk 4 above)
Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/data/source/local/dao/BookingDAOImpl.kt]: em nhầm chỗ này ạ

File: app/src/main/java/com/sunasterisk/travelapp/data/models/Location.kt [hunk 9]
```diff
@@ -69,11 +104,29 @@ data class Location(
...
+                            propertiesNumber =
+                                locationJSONObject
+                                    .getJSONObject(CATEGORY_COUNT)
+                                    .getJSONObject(RESTAURANTS)
+                                    .getString(TOTAL).toInt()
+                        }
+                        else -> {
+                            description = locationJSONObject.getString(DESCRIPTION)
+                        }
                     }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/models/Location.kt]: Nếu chỉ có 2 trường hợp thì dùng if else

**Pull request merged**